### PR TITLE
feat: add support for fluxomics in data overlay

### LINKED ISSUE #1066: reaction data overlay
**Please describe the feature**

As a user, I want to overlay reaction data, so that I can see which reactions have low scores and can be a bottleneck.

**Acceptance Criteria**

- [ ] view default data as reaction overlay (from the data-files repository)
- [ ] view user-uploaded data as reaction overlay
- [ ] The web page documentation is updated accordingly
- [ ] The changes are in alignment with the privacy policy
- [ ] The changes are tested on multiple models and browsers
- [ ] Tests verifying the changes are added
- [ ] The API response is the same as before
- [ ] The changes do not affect the deployment
- [ ] The changes do not reduce the performance

**Additional context**

The reaction data will follow the already defined standard, and will work similar to the transcriptomics data overlay, but color reaction elements instead of genes.
The data will be provided in the data-files repository.

### PR DESCRIPTION
**Pull request by MalinAhlberg**
**Related issue(s) and PR(s)**  
This PR closes #1066  and relates to private issue 138. Joint work with @e0 .

Adds functionality for data overlay for fluxomics.

**Type of change**  

- [X] New feature (non-breaking change which adds functionality)

 
**List of changes made**  

- added data type component in mapping in `frontend/src/helpers/dataOverlay.js`

**Screenshot of the fix**  

2022-08-15-141959_941x635_scrot

**Testing**  

- A certain branch in other repos is needed
- A rebuild is needed due to new dependencies
- Relative urls that can be reused both for production and local testing
- Instructions on how to test
1. Get the latest changes from [`data-files`](https://github.com/MetabolicAtlas/data-files/pull/31) and [`data-generation`](https://github.com/MetabolicAtlas/data-generation/pull/32).
2. Rebuild and restart the stack `build-stack && start-stack`.
3. Visit `/explore/Human-GEM/map-viewer/golgi_apparatus`. Expand the Data Overlay section. Verify that `fluxomics` is available under `Select data type`.
4. Select a level under `Levels from Human Protein Atlas 21.0`. Verify that the reactions in the map get re-colored.
5. Please also test on the `3D` map.
6. Upload custom expression data (example below) including reactions. Verify that the reactions in the map get re-colored.
7. Visit `http://localhost/documentation#data-overlay`. Verify that the documentation is updated and is correct.

`reaction1.mock.tsv`
```tsv
id	tissue 1	tissue 2
MAR01591	0.6	0.3
MAR00447	0.55	0.9
```

**Further comments**  

**Definition of Done checklist**  
- [X] My code meets the Acceptance Criteria
- [X] My code follows the [NBIS style guidelines](https://github.com/NBISweden/development-guidelines)
- [X] I have performed a self-review of my own code and commented any hard-to-understand areas
- [X] Tests and lint/format validations are passing
- [X] My changes generate no new warnings

### CODE REVIEW

**Comment by e0** [comment]
Just want to note this for reminder. After we integrate and test the new data files, we might want to update the `/documentation#data-overlay` section.

**Review by inghylt — CHANGES_REQUESTED** [review]
- Seems like the data overlay have become a lot slower? When I visit `/explore/Human-GEM/map-viewer/golgi_apparatus`, I can't really click any of the drop downs or any other map for that matter for noticeable period of time. If it's not possible to speed things up, I think we should add some kind of loading visualization 

- The slow speed is also noticeable when you switch to Interaction Partners, it hangs and behaves strange. I'm not sure if that is what we want to have, but I'm not noticing any coloring affect either if you right click a node and press one of the "Highlight reaction" options and use some fluxomics data

**Comment by e0** [comment]
Thanks for the great feedback @inghylt!

As discussed during the stand-up, we will try to split the data loading into list of dataset names and then dataset overlay info instead, as opposed to loading everything at once.

@mihai-sysbio Is it correct to assume that fluxomics should not be included in Interaction Partners? If not, please specify what should be done there.

**Comment by mihai-sysbio** [comment]
> we will try to split the data loading into list of dataset names and then dataset overlay info instead, as opposed to loading everything at once.

sounds like a very good approach, especially as we intend to add even more data

> Is it correct to assume that fluxomics should not be included in Interaction Partners?

tldr; that's correct. 

For the Map Viewer, the aim is to change visually the reaction nodes. Since these don't exist in the Interaction Partners, a substitute would be needed. There doesn't seem to be any good candidates for the _Interaction partners_.

**Review by nanjiangshu — COMMENTED** [review]
Great work @e0 and @MalinAhlberg. I tested the data overlay for reactions and all seem to work very well and I did not experience any slow down. Is the commit 872b763ed012d6f0cbf6236b522a6e6b86ecfc5d  that fixed the speed problem that @inghylt mentioned? The updates in the documentation looks well to me. A few minor comments that might not be address
1. Since most of the expression levels are 1, the color of highlights for expression levels (orange-brown) and the selected reaction (red) is hard to distinguish. Should we use a better color spectrum or choose a different color for selected item?
2. For Interaction Partners, only can upload the reaction mock file and select the tissue but nothing will happen, shall we show a warning/error message if the user uploaded data in Interaction Partners contains only reaction IDs, which could be uploaded by mistake.

**Comment by e0** [comment]
> Is the commit [872b763](https://github.com/MetabolicAtlas/MetabolicAtlas/commit/872b763ed012d6f0cbf6236b522a6e6b86ecfc5d) that fixed the speed problem that @inghylt mentioned?

This is the commit with most of the changes for the speed problem: https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1092/commits/39e468490ba69b5ee6b027742c7ccbc227792efa

___

> 1. Since most of the expression levels are 1, the color of highlights for expression levels (orange-brown) and the selected reaction (red) is hard to distinguish. Should we use a better color spectrum or choose a different color for selected item?
> 2. For Interaction Partners, only can upload the reaction mock file and select the tissue but nothing will happen, shall we show a warning/error message if the user uploaded data in Interaction Partners contains only reaction IDs, which could be uploaded by mistake.

I think these are good candidates for new issues. Personally I think they are low-priority for now but I would be happy to discuss more about them in the next backlog refinement.

**Comment by mihai-sysbio** [comment]
@nanjiangshu very good points. I can add a bit more.

> 1. a better color spectrum

There is a more general issue already reported - to adjust the color scheme for color blindness https://github.com/MetabolicAtlas/private-issues/issues/35 .

> 2\. but nothing will happen

There are even more corner cases: say one zooms in on a map, and uploads custom data that doesn't affect that part of the map. From my point of view it becomes the user's responsibility to "align" the data they have uploaded to the thing they are seeing. As long as the documentation clearly says that invalid IDs are ignored, which could be updated when documentation is added to this PR (expectation created by [this comment](https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1092#issuecomment-1214955929)).

**Review by mihai-sysbio — CHANGES_REQUESTED** [review]
Looks great! I'm also very happy to see the that the previous refactoring work of the data overlay is making it easier to implement this. Also, awesome improvement in the way of handling data, we do want this feature to not give the impression of a slow website.

I did notice confusing behaviour around the value highlighting:
- there seems to be no value being shown for reaction data, which is different behaviour than for the genes
- not sure if it's related to #973, but after selecting transcriptomics and reverting to fluxomics, I could do this:
image

File: frontend/src/components/Documentation.vue [hunk 1]
```diff
@@ -290,9 +290,7 @@
                 ). To clear the RNA levels, select the
                 <i>None</i>
                 option in the drop down list. RNA levels are available for both 2D and 3D Map
-                Viewer. The gene expression levels used are obtained from the file
-                <b>rna_tissue_hpa.tsv.zip</b>
-                in the
+                Viewer. The expression levels used are obtained from the
```
[the remark this answers is not in the document]
Comment (johan-gson) [reply frontend/src/components/Documentation.vue:293]: With the risk of adding confusion, I would say that what we are looking at for reactions is reaction presence. The number between 0-1 is really a number between 0 and 100, just scaled down. What we did here was to generate 100 bootstraps from the single-cell RNA-Seq data, pool those single-cells into RNA-Seq profiles, and then run ftINIT on each of those, which for each bootstrap determines the presence of each reaction. The number 0-100 represents in how many bootstraps a reaction was present, so the scaled value 0.5 means that the reaction was considered present in 50 out of 100 bootstrap models. Bootstrapping is a way to "randomize" the RNA-Seq data a bit - the purpose of this is to get a metric of how certain we are that a reaction is present. A value of >=0.99 thus represents that we are pretty sure it is present, <= 0.1 that it is not present, and for reactions with values in between we are really not sure if the reaction is present or not. So, the values have very little to do with fluxes. I hope this helps.

**Comment by e0** [comment]
Thanks for the careful reviewing and feedback @mihai-sysbio! The mouseover issue should be fixed now in: https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1092/commits/a6c2fbfbaafae27cc590a7f2aa36444fb9de5461

File: frontend/src/components/Documentation.vue [hunk 2]
```diff
@@ -280,7 +280,7 @@
                 <i>Map Viewer</i>. By clicking the
                 <i>Data overlay</i>
                 button on the right side of the page, one can toggle the hidden/display of this
-                panel. For Human-GEM, the gene expression levels for genes from
+                panel. For Human-GEM, the gene and reaction expression levels from
```
[the remark this answers is not in the document]
Comment (e0) [reply frontend/src/components/Documentation.vue:283]: How about we simply remove the word `expression` from here for now since correct naming is still undecided and could be addressed in a new issue?

File: frontend/src/components/Documentation.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (e0) [reply frontend/src/components/Documentation.vue:293]: Similar to [this comment](https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1092#discussion_r947634418), it would be nice if we can remove the word `expression` from here as well for now so that we can close this issue. It would be `The levels used are obtained from the ...`.

File: frontend/src/components/Documentation.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (mihai-sysbio) [reply frontend/src/components/Documentation.vue:283]: In that case, please commit here the more generic explanation:
```suggestion
                panel. For Human-GEM, the gene expression levels and reaction data from
```

File: frontend/src/components/Documentation.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (e0) [reply frontend/src/components/Documentation.vue:293]: I think the idea behind the original change (before this first suggestion) was to have something generic so it includes genes, reactions, and potentially other types we might have in the future. It seems a bit redundant to have the same sentence again and only switching out `gene` with `reaction`. I might be misunderstanding what you mean though and if that's the case please provide a full suggestion including what should be added for reaction.

[the remark this answers is not in the document]
Comment (mihai-sysbio) [reply frontend/src/components/Documentation.vue:293]: The original section details how the gene expression data was obtained (from where it was downloaded). An improvement would be to say how it was processed and link to the instructions in the `data-files` repository.

That last sentence, either with `expression levels` but better with `gene expression levels` is correct.

What should be added is a new paragraph - a concise version of Johan's explanation above, something going like this:

> The reaction presence was initially computed as a number between 0 and 100, scaled to 0-1 as per [these instructions](). The number is computed by creating bootstraps from the single-cell RNA-Seq data from the [Human Protein Atlas - Single-cell type Section](), pooling those into RNA-Seq profiles and applying the ftINIT algorithm. A high value >=0.9 means that there is high confidence that the reaction is present, whereas a low value <=0.1 means it is not. For more information on single-cell bootstraps and how ftINIT works, please see [this manuscript]().

[the remark this answers is not in the document]
Comment (e0) [reply frontend/src/components/Documentation.vue:293]: Thanks @mihai-sysbio. I found the link for the HPA Single-cell stuff. Could you please provide the links to the instructions and the manuscript?

**Comment by inghylt** [comment]
Everything seems to work well now! When the Documentation discussion is sorted I am redo to approve

File: frontend/src/components/Documentation.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (mihai-sysbio) [reply frontend/src/components/Documentation.vue:293]: The text above has been slightly tweaked and now includes all the links.

[the remark this answers is not in the document]
Comment (mihai-sysbio) [reply frontend/src/components/Documentation.vue:293]: @johan-gson could you please confirm if the thresholds are: (a) `0.99 and 0.01` or (b) `0.9 and 0.1`? Based on the manuscript I think it's (a).

[the remark this answers is not in the document]
Comment (e0) [reply frontend/src/components/Documentation.vue]: The initial suggestion from this thread as well as the new paragraph has been links have been addressed in: https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1092/commits/63dfa468105b81574bef6c24ef2c07f4c8536825

[the remark this answers is not in the document]
Comment (johan-gson) [reply frontend/src/components/Documentation.vue]: The threshold we used for on is >= 0.99 and off is <= 0.01

File: frontend/src/components/Documentation.vue [hunk 3]
```diff
@@ -290,14 +290,32 @@
...
                 of the Human Protein Atlas.
               </p>
+              <p>
+                The reaction presence was initially computed as a number between 0 and 100. The
+                value is computed by creating bootstraps from the single-cell RNA-Seq data from the
+                <a href="https://www.proteinatlas.org/about/releases#21.0" target="_blank"
+                  >Human Protein Atlas - Single Cell Type Section</a
+                >, pooling those into RNA-Seq profiles and applying the ftINIT algorithm. A high
+                value >=0.9 means that there is high confidence that the reaction is present,
+                whereas a low value <=0.1 means it is not. The resulting values were then scaled to
```
Comment (mihai-sysbio) [inline frontend/src/components/Documentation.vue:306]: The value needs to be corrected in light of @johan-gson's comment above, and the HTML character should be used:
```suggestion
                whereas a low value &#2264;=0.01 means it is not. The resulting values were then scaled to
```

File: frontend/src/components/Documentation.vue [hunk 4]
```diff
@@ -290,14 +290,32 @@
...
                 </a>
                 of the Human Protein Atlas.
               </p>
+              <p>
+                The reaction presence was initially computed as a number between 0 and 100. The
+                value is computed by creating bootstraps from the single-cell RNA-Seq data from the
+                <a href="https://www.proteinatlas.org/about/releases#21.0" target="_blank"
+                  >Human Protein Atlas - Single Cell Type Section</a
+                >, pooling those into RNA-Seq profiles and applying the ftINIT algorithm. A high
+                value >=0.9 means that there is high confidence that the reaction is present,
```
Comment (mihai-sysbio) [inline frontend/src/components/Documentation.vue:305]: The value needs to be corrected in light of @johan-gson's comment above, and the HTML character should be used:
```suggestion
                value &#8805;0.99 means that there is high confidence that the reaction is present,
```

**Review by mihai-sysbio — APPROVED** [review]
I think this is ready to be wrapped up! Great job seeing this one through!

There are 2 minor suggestions left, but they are straightforward hence my approval.

File: frontend/src/components/Documentation.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (mihai-sysbio) [reply frontend/src/components/Documentation.vue]: Suggestions have been made accordingly, so I'm resolving this thread.

File: frontend/src/components/Documentation.vue (hunk 3 above)
Comment (MalinAhlberg) [reply frontend/src/components/Documentation.vue]: Fixed in 1c9047e and b1a9f2b

File: frontend/src/components/Documentation.vue (hunk 4 above)
Comment (MalinAhlberg) [reply frontend/src/components/Documentation.vue]: Fixed in 1c9047e

**Review by inghylt — APPROVED** [review]

**Comment by nanjiangshu** [comment]
> @nanjiangshu very good points. I can add a bit more.
> 
> > 1. a better color spectrum
> 
> There is a more general issue already reported - to adjust the color scheme for color blindness [MetabolicAtlas/private-issues#35](https://github.com/MetabolicAtlas/private-issues/issues/35) .
> 
Great that there is already an issue for this problem
> > 2. but nothing will happen
> 
> There are even more corner cases: say one zooms in on a map, and uploads custom data that doesn't affect that part of the map. From my point of view it becomes the user's responsibility to "align" the data they have uploaded to the thing they are seeing. As long as the documentation clearly says that invalid IDs are ignored, which could be updated when documentation is added to this PR (expectation created by [this comment](https://github.com/MetabolicAtlas/MetabolicAtlas/pull/1092#issuecomment-1214955929)).

I agree this problem can be ignored if it is a very uncommon use case. Although I think it might be useful to let the user know that no ID in the custom data is found on the Map or the Interaction Partners so that they will not be confused that no highlighting will occur upon data overlay.

File: frontend/src/components/Documentation.vue [hunk 5]
```diff
@@ -303,7 +303,7 @@
                   >Human Protein Atlas - Single Cell Type Section</a
                 >, pooling those into RNA-Seq profiles and applying the ftINIT algorithm. A high
                 value &#8805;0.99 means that there is high confidence that the reaction is present,
-                whereas a low value &#8804;0.1 means it is not. The resulting values were then
+                whereas a low value &#8804;0.01 means it is not. The resulting values were then
```
[the remark this answers is not in the document]
Comment (mihai-sysbio) [reply frontend/src/components/Documentation.vue:306]: A complete explanation should not only cover that, but how the number is computed as well.
The way I see things at the moment, I propose to leave things as they are for now. If/when the manuscript gets published, the link would need to also be updated, which would make it a natural moment to also update this description accordingly.

**Comment by mihai-sysbio** [comment]
> It looks really good now. I only left a minor comment regarding the description of the reaction threshold, more precisely, for the values in between 0.01 and 0.99. I can approve this PR if this is solved and the documentation of the reaction part are verified by the researchers. 

You do make a good point @nanjiangshu. I will keep this in mind for the next data overlays to be added (one is foreseen to be coming up as early as next week).
In the meantime, I can already inform you that the researcher has seen the feature in a demo from me, and also has access to it on the dev server, and no changes were immediately required.
